File: Data Structures & Algorithms/best-time-to-buy-and-sell-stock-ii/submission-3.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        i, j = 0, 0
        res = 0
        # while (j<len(prices)):
        #     if (j+1)<len(prices) and prices[j+1]>=prices[j]:
        #         j+=1
        #     elif (j+1)<len(prices) and prices[j+1]<prices[j] and prices[i]<prices[j]:
        #         res+= (prices[j]-prices[i])
        #         i = j+1
        #         j = i+1
        #     elif (j+1)==len(prices):
        #         if prices[i]<prices[j]:
        #             res+= (prices[j]-prices[i])
        #         break
        #     else:
        #         j+=1
        while j<len(prices):
            if prices[i]<=prices[j]:
                if ((j+1)<len(prices) and prices[j+1]<prices[j]) or \
                   (j+1)==len(prices):
                    # get value
                    res+= prices[j]-prices[i]
                    i = j
                    j = i+1
                    continue
                j+=1
            else:
                i = j
                j = i+1

        return res

File: Data Structures & Algorithms/anagram-groups/submission-3.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        //先创建26长度的数组，统计每个字符串的字母
        //存入相应的hashmap中
        Map<String, List> map = new HashMap<>();
        for(String str: strs){
            int[] arr = new int[26];
            for (int i=0;i< str.length(); i++){
                arr[str.charAt(i)-'a']++;
            }
            //List arr2 = Arrays.asList(arr);
            StringBuffer sb = new StringBuffer();
            for (int x: arr){
                sb.append(x+"#");
            }
            String key = sb.toString();
            if(!map.containsKey(key)){
                List tmpList = new ArrayList<>();
                tmpList.add(str);
                map.put(key, tmpList);
            }else{
                List tmpList = map.get(key);
                tmpList.add(str);
                map.put(key, tmpList);
            }
        }
        List<List<String>> res = new ArrayList<>();
        for (List value: map.values()){
            res.add(value);
        }
        return res;
    }
}
